File: tools/generate-icons.js
/**
 * PWA图标生成脚本
 * 使用 Canvas 生成所有缺失的PWA图标
 */

const fs = require('fs');
const path = require('path');
const { createCanvas } = require('canvas');

// 需要生成的图标尺寸
const sizes = [72, 96, 128, 144, 152, 192, 384, 512];

// 输出目录
const outputDir = path.join(__dirname, '../assets/icons');

// 确保输出目录存在
if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
}

/**
 * 生成单个图标
 */
function generateIcon(size) {
    const canvas = createCanvas(size, size);
    const ctx = canvas.getContext('2d');

    // 背景渐变
    const gradient = ctx.createLinearGradient(0, 0, size, size);
    gradient.addColorStop(0, '#6666ff');
    gradient.addColorStop(1, '#4444cc');
    ctx.fillStyle = gradient;
    ctx.fillRect(0, 0, size, size);

    // 绘制声波图案
    ctx.strokeStyle = 'rgba(255, 255, 255, 0.3)';
    ctx.lineWidth = size / 40;
    ctx.lineCap = 'round';

    for (let i = 0; i < 5; i++) {
        ctx.beginPath();
        const amplitude = size / 8;
        const frequency = 0.05;
        const yOffset = size / 2 + (i - 2) * (size / 10);

        for (let x = 0; x <= size; x += 2) {
            const y = yOffset + Math.sin(x * frequency + i) * amplitude;
            if (x === 0) {
                ctx.moveTo(x, y);
            } else {
                ctx.lineTo(x, y);
            }
        }
        ctx.stroke();
    }

    // 中心圆形（代表声音源）
    const centerX = size / 2;
    const centerY = size / 2;
    const circleRadius = size / 8;

    ctx.fillStyle = 'rgba(255, 255, 255, 0.9)';
    ctx.beginPath();
    ctx.arc(centerX, centerY, circleRadius, 0, Math.PI * 2);
    ctx.fill();

    // 添加文字（仅在较大尺寸）
    if (size >= 192) {
        ctx.fillStyle = '#6666ff';
        ctx.font = `bold ${size / 10}px Arial`;
        ctx.textAlign = 'center';
        ctx.textBaseline = 'middle';
        ctx.fillText('SF', centerX, centerY);
    }

    return canvas;
}

/**
 * 保存图标为PNG文件
 */
function saveIcon(canvas, size) {
    const filename = `icon-${size}x${size}.png`;
    const filepath = path.join(outputDir, filename);
    const buffer = canvas.toBuffer('image/png');
    fs.writeFileSync(filepath, buffer);
    console.log(`✅ 生成: ${filename}`);
}

/**
 * 生成所有图标
 */
function generateAllIcons() {
    console.log('🎨 开始生成PWA图标...\n');

    sizes.forEach(size => {
        const canvas = generateIcon(size);
        saveIcon(canvas, size);
    });

    // 额外生成 favicon.png 和 android-chrome-192x192.png
    const favicon32 = generateIcon(32);
    fs.writeFileSync(path.join(__dirname, '../favicon.png'), favicon32.toBuffer('image/png'));
    console.log('✅ 生成: favicon.png (32x32)');

    const androidChrome = generateIcon(192);
    fs.writeFileSync(path.join(__dirname, '../android-chrome-192x192.png'), androidChrome.toBuffer('image/png'));
    console.log('✅ 生成: android-chrome-192x192.png');

    console.log('\n✅ 所有图标生成完成！');
    console.log(`📁 PWA图标位置: ${outputDir}`);
    console.log('📁 根目录图标: favicon.png, android-chrome-192x192.png');
}

// 执行生成
try {
    generateAllIcons();
} catch (error) {
    console.error('❌ 生成图标失败:', error.message);
    console.log('\n💡 提示: 请先安装 canvas 包:');
    console.log('   npm install canvas');
    process.exit(1);
}

File: tools/generate-icons.py
#!/usr/bin/env python3
"""
PWA图标生成脚本
使用 PIL 生成所有缺失的PWA图标
"""

import os
import math
from PIL import Image, ImageDraw, ImageFont

# 需要生成的图标尺寸
sizes = [72, 96, 128, 144, 152, 192, 384, 512]

# 输出目录
script_dir = os.path.dirname(os.path.abspath(__file__))
project_dir = os.path.dirname(script_dir)
icons_dir = os.path.join(project_dir, 'assets', 'icons')

# 确保输出目录存在
os.makedirs(icons_dir, exist_ok=True)

def generate_icon(size):
    """生成单个图标"""
    # 创建图像
    img = Image.new('RGB', (size, size), color='#0d131f')
    draw = ImageDraw.Draw(img)

    # 绘制背景渐变 (简化版：使用单色背景和圆形)
    # 主背景色
    draw.rectangle([(0, 0), (size, size)], fill='#6666ff')

    # 绘制声波效果（简化版：多个圆环）
    center_x = size // 2
    center_y = size // 2

    # 绘制多个同心圆环表示声波
    for i in range(1, 5):
        radius = int(size * 0.1 * i)
        alpha = max(10, 80 - i * 15)  # 逐渐变淡
        color = f'rgba(255, 255, 255, {alpha/255})'
        # PIL不直接支持rgba字符串，改用RGB
        rgb_color = (255, 255, 255)
        draw.ellipse(
            [(center_x - radius, center_y - radius),
             (center_x + radius, center_y + radius)],
            outline=rgb_color,
            width=max(1, size // 50)
        )

    # 中心实心圆（代表声音源）
    circle_radius = size // 8
    draw.ellipse(
        [(center_x - circle_radius, center_y - circle_radius),
         (center_x + circle_radius, center_y + circle_radius)],
        fill=(255, 255, 255)
    )

    # 添加文字（仅在较大尺寸）
    if size >= 192:
        try:
            # 尝试使用系统字体
            font_size = size // 10
            try:
                font = ImageFont.truetype("arial.ttf", font_size)
            except:
                font = ImageFont.load_default()

            text = "SF"
            # 获取文本边界框
            bbox = draw.textbbox((0, 0), text, font=font)
            text_width = bbox[2] - bbox[0]
            text_height = bbox[3] - bbox[1]

            # 在中心绘制文字
            text_x = center_x - text_width // 2
            text_y = center_y - text_height // 2
            draw.text((text_x, text_y), text, fill=(102, 102, 255), font=font)
        except Exception as e:
            print(f"   [WARN] Text rendering failed: {e}")

    return img

def generate_all_icons():
    """生成所有图标"""
    print(">>> Starting PWA icon generation...\n")

    # 生成各种尺寸的图标
    for size in sizes:
        img = generate_icon(size)
        filename = f'icon-{size}x{size}.png'
        filepath = os.path.join(icons_dir, filename)
        img.save(filepath, 'PNG')
        print(f"[OK] Generated: {filename}")

    # 生成 favicon.png (32x32)
    favicon = generate_icon(32)
    favicon_path = os.path.join(project_dir, 'favicon.png')
    favicon.save(favicon_path, 'PNG')
    print(f"[OK] Generated: favicon.png (32x32)")

    # 生成 android-chrome-192x192.png
    android_chrome = generate_icon(192)
    android_path = os.path.join(project_dir, 'android-chrome-192x192.png')
    android_chrome.save(android_path, 'PNG')
    print(f"[OK] Generated: android-chrome-192x192.png")

    print(f"\n[SUCCESS] All icons generated!")
    print(f"[INFO] PWA icons location: {icons_dir}")
    print(f"[INFO] Root icons: favicon.png, android-chrome-192x192.png")

if __name__ == "__main__":
    try:
        generate_all_icons()
    except Exception as error:
        print(f"[ERROR] Icon generation failed: {error}")
        print("\n[TIP] Please install Pillow package:")
        print("   pip install Pillow")
        exit(1)
